src/ac_power_flow_jacobian.jl

```diff
@@ -1,228 +1,511 @@
-                    #Jac: Reactive PF against same angle θ[ix_f]
-                    J[F_ix_f_i, X_ix_t_snd] =
-                        Vm[ix_f] * sum(
-                            Vm[k] * (
-                                real(Yb[ix_f, k]) * cos(θ[ix_f] - θ[k]) -
-                                imag(Yb[ix_f, k]) * -sin(θ[ix_f] - θ[k])
-                            ) for k in data.neighbors[ix_f] if k != ix_f
-                        )
-                else
-                    g_ij = real(Yb[ix_f, ix_t])
-                    b_ij = imag(Yb[ix_f, ix_t])
-                    #Jac: Active PF w/r to different voltage magnitude Vm[ix_t]
-                    J[F_ix_f_r, X_ix_t_fst] =
-                        Vm[ix_f] *
-                        (g_ij * cos(θ[ix_f] - θ[ix_t]) + b_ij * sin(θ[ix_f] - θ[ix_t]))
-                    #Jac: Active PF w/r to different angle θ[ix_t]
-                    J[F_ix_f_r, X_ix_t_snd] =
-                        Vm[ix_f] *
-                        Vm[ix_t] *
-                        (g_ij * sin(θ[ix_f] - θ[ix_t]) + b_ij * -cos(θ[ix_f] - θ[ix_t]))
-
-                    #Jac: Reactive PF w/r to different voltage magnitude Vm[ix_t]
-                    J[F_ix_f_i, X_ix_t_fst] =
-                        Vm[ix_f] *
-                        (g_ij * sin(θ[ix_f] - θ[ix_t]) - b_ij * cos(θ[ix_f] - θ[ix_t]))
-                    #Jac: Reactive PF w/r to different angle θ[ix_t]
-                    J[F_ix_f_i, X_ix_t_snd] =
-                        Vm[ix_f] *
-                        Vm[ix_t] *
-                        (g_ij * -cos(θ[ix_f] - θ[ix_t]) - b_ij * sin(θ[ix_f] - θ[ix_t]))
-                end
-            else
-                error("Undefined Conditional")
-            end
+    for bus_from in 1:num_buses
+        row_from_p = 2 * bus_from - 1
+        row_from_q = 2 * bus_from
+
+        for bus_to in data.neighbors[bus_from]
+            col_to_vm = 2 * bus_to - 1
+            col_to_va = 2 * bus_to
+            bus_type = data.bus_type[bus_to, time_step]
+            _set_entries_for_neighbor(Jv, Yb, Vm, θ,
+                bus_from, bus_to,
+                row_from_p, row_from_q,
+                col_to_vm, col_to_va,
+                data.neighbors[bus_from], Val(bus_type))
         end
     end
     return
 end
+
+function calculate_loss_factors(
+    data::ACPowerFlowData,
+    Jv::SparseMatrixCSC{Float64, Int32},
+    time_step::Int,
+)
+    num_buses = first(size(data.bus_type))
+    ref, pv, pq = bus_type_idx(data, time_step)
+    pvpq = vcat(pv, pq)
+    pvpq_coords = [
+        x for pair in zip(
+            [2 * x - 1 for x in 1:num_buses if x in pvpq],
+            [2 * x for x in 1:num_buses if x in pvpq],
+        ) for x in pair
+    ]
+    data.loss_factors[ref, time_step] .= 1.0
+    penalty_factors!(
+        Jv[pvpq_coords, pvpq_coords],
+        vec(collect(Jv[2 .* ref .- 1, pvpq_coords])),
+        view(
+            data.loss_factors,
+            [x for x in 1:num_buses if x in pvpq],
```

Comment:
Wait what is this even doing? Making a list of numbers that are in `pvpq`? Isn't that just `pvpq`? Do we need to sort them or something?

Reply:
yes sorting it is much easier.